Implement proposed database schema

bcipolli opened the pull request:
This also works off of #101. This implements the database schema proposed here:
https://github.com/caciviclab/caciviclab.github.io/wiki/Data-relationships---Database-schema

With a full schema such as this, we can begin to build the APIs required for the front-end apps, and start to fill the database with relevant data from `NetFile` and `CalAccess`.

Besides implementing the schema above, this PR:
- Renames `netfile` to `netfile_raw`. The idea is to have a stand-alone app that simply loads raw netfile data into a database table, so that data munging can happen in Python/Django rather than through csv/pandas.
- Moves all of `zipcode_metro` into `locality`. Note that the management command for loading zip codes is temporarily broken.

To deal with lacking a globally unique ID, I've pushed the `Locality` model to contain such an ID. It'll be an arbitrary ID that can be used, app-wide, to refer to things. I don't know if it'll ever be needed, but it's in the doc :)

@jnmarcus Really gotta run, but wanted to push this up for you & others to consider! May be hard to read without #101 merged, however...

tdooner commented:
What do all the things you start you commit messages mean?

bcipolli commented:
Oh, perhaps this is just for the neuroimaging in Python community? I thought it's standard use, but the only docs when I searched were here:
http://nipy.org/nipy/devel/guidelines/commit_codes.html

tdooner commented:
Interesting, I've never seen it before, but also I'm not involved at all in the python community so that probably explains it.

adborden commented on ballot/migrations/0002_auto_20151219_0359.py:
```diff
@@ -7,7 +7,7 @@
 
     dependencies = [
-        ('ballot', '0002_load_ballot_data'),
+        ('ballot', '0001_initial'),
```
Oh, this is clever. At first I was thinking you were going to break the migrations, but I guess since the load_ballot_data wasn't making any schema changes, it's safe to just remove that and patch the link.

adborden commented:
@bcipolli thanks for doing this work! I have some comments on the Data
relationships schema wiki, but I guess I'll leave them here? I don't think
editing the wiki would help much.

Github doesn't do threading, so I wonder if tackling too many topics on the PR
discussion is going to get out of hand. If that's the case, maybe opening up
issues for individual topics might be a better approach, and we'll open up
issues as folks bring them up here. If anyone else has any ideas, pitch away.

## 

## Model inspiration

I'm curious where you got the inspiration for the fields on some of these
models. For example, the `Person` model with first, middle, last, party. There's
a lot more fields in the netfile data (just looking at the columns) and party
isn't one of them. I'd prefer we model off of some existing work, or something
based on the data we have.

FWIW, curious if @tdooner or @mikeubell have insights from the OpenOakland
Disclosure project. I would use that as a starting point for the contribution
models.

## Committee as Contributor?

I'm reading this as: a Contributor is one of Individual, Company, or Committee,
but I don't think this is correct.

Committees exist to support/oppose both Ballot Measures and Candidates, as
a ballot committee or a candidate committee, respectively.  I believe they
receive contributions from others, and spend money based on their declared
stance. I don't think they are contributors, because they don't contribute money
to a particular issue or candidate, they only spend on behalf of that issue or
candidate.

Does the Committee represent something different?

## Contribution app as catchall

I'd like to spend a little more time thinking about how to break down the
financial domain. We recognize the finance part of this project being the most
complicated and the most unknown, so it doesn't feel right to shove it all in
a single app. (Of course, we have to start some where!).

From your proposal, I see two main domains here, the Ballot domain (what
issues/candidates am I voting for) and the Finance domain (where is the money
coming and going).  There could be other domains as well, we just haven't picked
up on them yet.

_Side note, I think Locality itself, might be yet another domain, since each Locality
has it's own rules, laws, and filing dates, even though a lot is dictated by the
State of California. This data doesn't seem like it belongs within the Ballot
domain, hence it smells like something new._

This is a picture I would draw from looking at your PR (it's incomplete, only
using it here to illustrate the domain concept).

![domain-venn-diagram](https://cloud.githubusercontent.com/assets/509703/12047452/8d8becca-ae80-11e5-9a6a-caadf3fbcf8a.png)

I would think that `ballot` and `candidate` are not strictly in the Ballot
domain, that each might know something about the money and the Finance around
these entities. Maybe these would be the `ballot_finance` and
`candidate_finacnce` apps.

### Three tiers

Although we don't have to be strict about this, I think it would be a helpful
exercise if we tried to constrain ourselves by thinking in these tiers (at
least, I found it useful).

In my head, I see three tiers. Low-level apps related to one of the two domains,
each likely being exclusive to a single domain (although not a hard rule). These
are the basic building blocks of our data. e.g. Elections, Ballot information,
Contributions.

The second tier, building on top of the first, starting to blur the lines
between the domains. Using the building blocks from the first tier to build
something new. e.g. `ballot_finance` (Contributions sliced with a Ballot focus),
`candidate_finance` (Contributions sliced with a Candidate focus), `locality_finance`
(Contributions aggregated at a Locality level).

The third tier, `disclosure` app, building on top of the previous ones.
Potentially a lobbying app would end up in this tier as well, but it's really
just the disclosure app for now.

_Sorry, no diagram :P_

_Forgive me if I'm digging too deep on this, we don't have to have the
conceptual model complete locked down at this point._

## Elections and other models being usurped

Kind of a minor thing, maybe I'd feel differently if I better understood where
these were going, but I feel like some of the models should exist, but they were
moved from one place or removed altogether.

Election as an election day - I think this is an important concept. Useful to
list out the election dates across the state and trace out historic data.
Doesn't feel right to fold it into the Ballot model.

Contest vs. Ballot Measure - Ballot Measures as a vocabulary word is already
a bit contentious because it goes by so many names, but now we're introducing the idea that a Ballot Meausre

I didn't spend too much time on thinking up names for models, but I'd rather
take a little effort to avoid "hotspots" of ambiguity.

## Admin

Just something to keep in mind. I'd rather not break the form input for folks
since we need it to get the app off the ground. The folks doing the entry aren't
going to be technical or campaign finance experts, so it's important the quality
of the admin site is high.

Thanks again, Ben!

bcipolli commented:
Hey @aborden thanks for sharing thoughts (and diagrams!) ahead of tonight. A lot of good stuff in there. Don't want to be too long, so just a few notes before tonight:
- I completely agree that we need to care about not just back-end storage and front-end APIs, but also data entry ease for non-tech people. Thanks for bringing up!
- I also agree that terms are crucial. I am pretty term-agnostic and still learning, so sorry for the mess. I do have a clear idea on the concepts I'm targeting, but terms may vary .. :)
- Some items that can be explicitly represented I've opted to have available as views/queries/API endpoints rather as explicit table names (e.g. Election as a query on top of Ballot). I totally agree that making the data available via API is crucial, but maybe differ on what the right storage on the back-end is.

I've suggested a few generalizations (all contributions together, ballot measure/ballot measure option as a way to represent core ballot measures and candidates, for common connectivity to contributions) that I don't think came across well in the doc, but that are pretty crucial to the approach. Perhaps I can try to flesh that out better tonight.

I also think there is not much separation between ballot measure/candidate and finance--each contribution is tied directly to a ballot measure choice (i.e. a canididate or reply to any ballot measure). It is essentially the `ballot_finance` app you mentioned.

As for other issues, I _think_ they can wait ... if not, lemme know and I can conjure up some magic to think about before tonight :)

Thanks again, and looking forward to discussing.

mikeubell commented:
A committee can be a contributor to a candidate (committee).  A committee
is just a registered entity that raises, spends and/or distributes money
for political purposes.
On Dec 29, 2015 11:08 PM, "Aaron D Borden" notifications@github.com wrote:

> @bcipolli https://github.com/bcipolli thanks for doing this work! I
> have some comments on the Data
> relationships schema wiki, but I guess I'll leave them here? I don't think
> editing the wiki would help much.
> 
> Github doesn't do threading, so I wonder if tackling too many topics on
> the PR
> discussion is going to get out of hand. If that's the case, maybe opening
> up
> issues for individual topics might be a better approach, and we'll open up
> issues as folks bring them up here. If anyone else has any ideas, pitch
> away.
> Model inspiration
> 
> I'm curious where you got the inspiration for the fields on some of these
> models. For example, the Person model with first, middle, last, party.
> There's
> a lot more fields in the netfile data (just looking at the columns) and
> party
> isn't one of them. I'd prefer we model off of some existing work, or
> something
> based on the data we have.
> 
> FWIW, curious if @tdooner https://github.com/tdooner or @mikeubell
> https://github.com/mikeubell have insights from the OpenOakland
> Disclosure project. I would use that as a starting point for the
> contribution
> models.
> Committee as Contributor?
> 
> I'm reading this as: a Contributor is one of Individual, Company, or
> Committee,
> but I don't think this is correct.
> 
> Committees exist to support/oppose both Ballot Measures and Candidates, as
> a ballot committee or a candidate committee, respectively. I believe they
> receive contributions from others, and spend money based on their declared
> stance. I don't think they are contributors, because they don't contribute
> money
> to a particular issue or candidate, they only spend on behalf of that
> issue or
> candidate.
> 
> Does the Committee represent something different?
> Contribution app as catchall
> 
> I'd like to spend a little more time thinking about how to break down the
> financial domain. We recognize the finance part of this project being the
> most
> complicated and the most unknown, so it doesn't feel right to shove it all
> in
> a single app. (Of course, we have to start some where!).
> 
> From your proposal, I see two main domains here, the Ballot domain (what
> issues/candidates am I voting for) and the Finance domain (where is the
> money
> coming and going). There could be other domains as well, we just haven't
> picked
> up on them yet.
> 
> _Side note, I think Locality itself, might be yet another domain, since
> each Locality has it's own rules, laws, and filing dates, even though a lot
> is dictated by the State of California. This data doesn't seem like it
> belongs within the Ballot domain, hence it smells like something new._
> 
> This is a picture I would draw from looking at your PR (it's incomplete,
> only
> using it here to illustrate the domain concept).
> 
> [image: domain-venn-diagram]
> https://cloud.githubusercontent.com/assets/509703/12047452/8d8becca-ae80-11e5-9a6a-caadf3fbcf8a.png
> 
> I would think that ballot and candidate are not strictly in the Ballot
> domain, that each might know something about the money and the Finance
> around
> these entities. Maybe these would be the ballot_finance and
> candidate_finacnce apps.
> Three tiers
> 
> Although we don't have to be strict about this, I think it would be a
> helpful
> exercise if we tried to constrain ourselves by thinking in these tiers (at
> least, I found it useful).
> 
> In my head, I see three tiers. Low-level apps related to one of the two
> domains,
> each likely being exclusive to a single domain (although not a hard rule).
> These
> are the basic building blocks of our data. e.g. Elections, Ballot
> information,
> Contributions.
> 
> The second tier, building on top of the first, starting to blur the lines
> between the domains. Using the building blocks from the first tier to build
> something new. e.g. ballot_finance (Contributions sliced with a Ballot
> focus),
> candidate_finance (Contributions sliced with a Candidate focus),
> locality_finance
> (Contributions aggregated at a Locality level).
> 
> The third tier, disclosure app, building on top of the previous ones.
> Potentially a lobbying app would end up in this tier as well, but it's
> really
> just the disclosure app for now.
> 
> _Sorry, no diagram :P_
> 
> _Forgive me if I'm digging too deep on this, we don't have to have the
> conceptual model complete locked down at this point._
> Elections and other models being usurped
> 
> Kind of a minor thing, maybe I'd feel differently if I better understood
> where
> these were going, but I feel like some of the models should exist, but
> they were
> moved from one place or removed altogether.
> 
> Election as an election day - I think this is an important concept. Useful
> to
> list out the election dates across the state and trace out historic data.
> Doesn't feel right to fold it into the Ballot model.
> 
> Contest vs. Ballot Measure - Ballot Measures as a vocabulary word is
> already
> a bit contentious because it goes by so many names, but now we're
> introducing the idea that a Ballot Meausre
> 
> I didn't spend too much time on thinking up names for models, but I'd
> rather
> take a little effort to avoid "hotspots" of ambiguity.
> Admin
> 
> Just something to keep in mind. I'd rather not break the form input for
> folks
> since we need it to get the app off the ground. The folks doing the entry
> aren't
> going to be technical or campaign finance experts, so it's important the
> quality
> of the admin site is high.
> 
> Thanks again, Ben!
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/caciviclab/disclosure-backend/pull/103#issuecomment-167949270
> .

adborden commented:
> A committee can be a contributor to a candidate (committee).  A committee
> is just a registered entity that raises, spends and/or distributes money
> for political purposes.

Good point, thanks @mikeubell

adborden commented:
@bcipolli sounds like we're on the same page. I might ask to tweak the names in the final PR. Basing the models on an existing work or someone else's paradigm is a good way to land on meaningful terms without us having to discuss it too much.

On the data representation, I certainly am thinking model centric as opposed to views/queries/API endpoints. Mostly because swagger actually [provides model declarations](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/1.2.md#526-models-object) for their API endpoints. We can describe these models in the yaml docstring but django-rest-swagger will actually do it for you if your view is based on model serializers, and hence django models.

BUT I am blurring your definition of django models in the PR with the conceptual models from the [Wiki](https://github.com/caciviclab/caciviclab.github.io/wiki/Data-relationships---Database-schema). So I suppose my comments are more about arguing to have these entities as first-class conceptual models in our schema design, even if though the schema itself is implemented as any of django model, serializer, or yaml string.

bcipolli commented:
@adborden Sounds great. I completely agree re: terms; in fact, I added a few to the Wiki glossary. It would be great to nail down those terms. Perhaps we can open an issue to work on that, since wikis seem like a poor place for discussion? It'd be great to nail terms down, and then practice using them.

Also, thanks for the pointer on the REST framework and its interaction with Swagger. I will take a look at how all that works, and let you know.

> We can describe these models in the yaml docstring but django-rest-swagger will actually do it for you if your view is based on model serializers, and hence django models.

I agree that using docstring YAML isn't desirable. However, defining serializers that expose properties of models and their foreign key references sounds like a clean way to go, without necessarily defining explicit models for each API endpoint.

> I suppose my comments are more about arguing to have these entities as first-class conceptual models in our schema design, even if though the schema itself is implemented as any of django model, serializer, or yaml string.

OK, cool, that definitely simplifies the discussion.

Can you lemme know which models you're most concerned with having as "1st class citizens"? I'm glad to discuss. My only goal is to define exactly as much as we need (nothing more, nothing less), have very simple dependencies between apps, and make each app stand-alone in a meaningful way (i.e. that conceptually, could be a building block in another project).

Commit:
ENH: branch locality app from ballot

Commit:
RF: split ballot app into ballot_measure and election_day apps.

Commit:
ENH: add candidate app

Commit:
ENH: add finance app

bcipolli commented:
OK, after working through some of the Netfile data, working on top of this schema for more than a week, and feedback from others, I've changed the structure a bit. Now, the apps (ordered by linear dependency):
- `locality` - deals with places. Main model is `Locality`, which is what all government-related tables should use to reference locations.
- `ballot_measure` - This contains everything related to individual ballot measures (`BallotMeasure` and `BallotMeasureChoice`). It also contains the `Ballot` model, which is just a group of ballot measures.
- `candidate` - Candidates are `BallotMeasureChoice`s. This app contains the metadata and restrictions that are special to candidates.
- `election_day` - This takes the `Ballot`, and connects event-related metadata. The main model is `ElectionDay`, to avoid conflict with `candidate.Election` - a "contest" for a particular government position in a particular locality.
- `finance` - This was called `contributions` before. However, this will contain data for contributions and independent expenditures, so I renamed it.
- `disclosure` - This is still the main app. It contains feeds that span `finance`, `candidate`, and `ballot_measure`.

I'm actually much happier with this structure than the previous one.

I've removed `WIP` as we're looking to get this merged in the next few days. @adborden Please have at it :)

bcipolli commented:
Eh... tests are failing because I fudged up some of the refactoring. Fixing now.

bcipolli commented:
OK @adborden @tdooner all set for review!

bcipolli commented:
Erm, worked on importing netfile data a bit last night, and that suggested a few changes. Will update here within an hour.

bcipolli commented:
OK, I have verified that this version can be used to _actually_ create relevant data from netfile, so this is a good one to review. @mrmaksimize we'll need to coordinate; I've done some preliminary work on `xform-netfile` to validate this schema.

adborden commented on ballot_measure/models.py:
```diff
@@ -0,0 +1,68 @@
+
+
+@python_2_unicode_compatible
+class BallotMeasureChoice(models.Model):
```
I think the term "ballot measure" is associated with a referendum or proposition. Using it to mean any item on the ballot doesn't feel right to me. I think I was calling this "contest" before, but I'm open to other ideas. Even "ballot item" would be preferred.

I do like the "choice" part though. "contest choice", "ballot item choice" or something.

adborden replied:
Maybe I'm confused here. I'm reading that a `BallotMeasure` is "A single referendum or candidate office which appears on a voter's Ballot." and can be an Office or a Referendum. One or more Candidates run for Office, but a Referendum is only Pass/Fail. So your BallotMeasureOptions would be different depending on whether the BallotMeasure is a Referendum or an Office. However, the fields on these models don't seem to reflect that.

Here's what I'm reading:

**If BallotMeasure is Office**

e.g. Mayor, you would have a BallotMeasureOption for each candidate running.

**If a BallotMeasureOption is Referendum**

e.g. Prop A, you would have a BallotMeasureOption for "Support" and an option for "Against".

Is this how I should be reading these models?

bcipolli replied:
Right. Yes!

tdooner replied:
:+1: a bit unintuitive, but I think this sounds like a pretty good way to model the options for an election.

adborden replied:
@bcipolli but what I don't understand, if a Candidate running for office is a `BallotMeasureChoice` how would these fields be used? They seem only relevant to a Referendum:
- title
- subtitle
- brief
- full_text

bcipolli replied:
OK. Lemme make some of the changes we talked about, and let's go from there.

adborden commented on ballot_measure/models.py:
```diff
@@ -0,0 +1,68 @@
+    con_statement = models.TextField(blank=True)
+
+    def __str__(self):
+        return 'Option: %s%s' % (self.title, self.subtitle)
```
s/Option/Choice/

bcipolli replied:
:+1:

adborden commented on candidate/models.py:
```diff
@@ -0,0 +1,93 @@
+
+
+@python_2_unicode_compatible
+class Election(models.Model):
```
I'd prefer not to use the word "Election" to specifically mean the election of a person. I was trying to use the word Election to be the event when citizens vote.

I would suggest using the name "OfficeElection", "ElectionOffice" or "CandidateOffice", because it's very much about an Office up for election. However, I'm a little unclear on what the distinction is between this model and a BallotMeasure, since I'm under the impression an office up for election would be a BallotMeasure.

adborden replied:
Could this just be merged with the Office model?

bcipolli replied:
> I was trying to use the word Election to be the event when citizens vote.

I think `Election` is used both for single office (that's the main Wikipedia entry) and the overall event. I'm not clear why you prefer to keep `Election` for the event, rather than the single office contest. From the reading I did online, I felt the stronger meaning was for the single office contest.

These vocabulary will be used throughout the app, including the front-end, and keeping things consistent across all levels will make things run more smoothly. Perhaps we can get advice from others for these terms?

As for `Office` vs. `Election` (as I'm using it here), I would prefer to keep them separate. `Office` exists independent of election or not. It's like a generic government object. I've used that same type of abstraction in other places, for example `Person` vs. `IndividualBenefactor`, etc. and I think separating static metadata (office, person) from functional metadata (election, finance) is a nice, consistent way to keep our models focused and as light as possible.

bcipolli replied:
> I think the term "ballot measure" is associated with a referendum or proposition. Using it to mean any item on the ballot doesn't feel right to me.  I think I was calling this "contest" before, but I'm open to other ideas. Even "ballot item" would be preferred.

Good point. I like `BallotItem`. I found `Contest` confusing as it doesn't obviously relate to `Ballot`. `Ballot`, `BallotItem`, `BallotItemChoice` makes the relationships pretty straightforward for me.. I'd like to go that route!

tdooner replied:
Yeah, I'm not sure about this either. On one hand, I don't really see the benefit of normalizing the event of election day into an `Election` object. I don't think we'll need to build any features that rely on that normalization, such as a list of all elections happening November 6th, 2016. And, it will just result in us having to join more tables in order to render a `Locality`.

On the other hand, this implementation of `Election` as a join table between `Office` and `BallotMeasure` is not exactly what I would expect when thinking of an election either. I think I like `OfficeElection` as an alternative name, or maybe a `BallotMeasure` can just relate directly to an `Office` (optionally, of course. Referendums have no office).

adborden replied:
> As for Office vs. Election (as I'm using it here), I would prefer to keep them separate. Office exists independent of election or not. It's like a generic government object.

That makes sense for Office. I'm fine to have an Office entity representing something outside of the election. However, if Election means "an office up for election on a ballot", what term would you propose for the election day event?

The Voter Information Project is writing a [spec](http://vip-specification.readthedocs.org/en/vip5/xml/index.html#elements) to standardize ballot information across the country. IMHO, their terminology avoids a lot ambiguity which is why I would prefer to standardize on their terms (not necessarily their spec). At the very least, I found reading through it helpful to think about the domain space more. Like how they associate Elections with different authorities (city, county, state, federal), which made sense to me which sounded relevant with what we're doing.

VIP is also representing [additional data](http://vip-specification.readthedocs.org/en/vip5/xml/elements/election.html) for their `Election` entity, like election type (city, state, federal) which I think will be relevant to how the contributions are reported (speculating).

adborden commented on finance/models.py:
```diff
@@ -0,0 +1,145 @@
+    name = models.CharField(max_length=255)
+    address = models.ForeignKey('locality.Address', null=True)
+    locality = models.ForeignKey('locality.Locality', null=True)
+    committee_id = models.CharField(max_length=16, unique=True)
```
I might name this something different than committee_id. I'm assuming this Id is coming from a third-party (netfile or something). But I also see us using the name `committee_id` on other models to refer to Committee.id, which might be confusing.

Is there another name the source has given it? Filer id?

bcipolli replied:
Good point. `filer_id` could be slightly confusing, as it doesn't match NetFile's `filerId` field exactly, but I think it's a good choice. Thanks!

adborden commented on finance/models.py:
```diff
@@ -0,0 +1,145 @@
+
+
+@python_2_unicode_compatible
+class Company(SocialMediaModel):
```
Can we name this "Corporation"? Company is a bit ambiguous (which might be the desire) but Corporation refers to a legal entity which is not a human being (at least in my mind).

bcipolli replied:
:+1:

tdooner replied:
I'm not sure the product need for this model. I can imagine us building richer pages for corporations, but treating all contributors identically is all we're trying to do at the beginning, right?

adborden commented on finance/models.py:
```diff
@@ -0,0 +1,145 @@
+        ('SA', 'Semi-annual'),
+        ('QU', 'Quarterly'),
+        ('OT', 'Other')
+    )
```
Can jurisdictions have different filing frequencies than listed? Just curious.

bcipolli replied:
> Can jurisdictions have different filing frequencies than listed? Just curious.

Not to my knowledge, after talking with @andrell81 and Stacey. I do think migrating later wouldn't be too troublesome, so I'm also not too worried about it. But if @andrell81 says otherwise, glad to design it in now.

adborden commented on locality/models.py:
```diff
@@ -0,0 +1,109 @@
+        help_text="the precinct's number e.g., 32 or 32A "
+                  "(alpha characters are legal)."
+    )
+    zip_code = models.ForeignKey('ZipCode')
```
This is probably a many-to-many relationship because of how zipcodes and precincts overlap?

bcipolli replied:
I really don't know, but probably!

Broadening the topic: many of the models in `locality` aren't used. They were more for concept of what could be done, rather than being useful at the moment. Probably best to comment out, or remove, such models.

adborden replied:
I think it's okay to leave the extras for now, and cull them as the concepts become more clear.

adborden commented:
Thanks @bcipolli, this is great! Read my comments more of "areas of clarification" rather than things that need code changes. I don't want to hold this PR up, but I do want to make sure there's minimal ambiguity for everyone about what all these apps and models represent even if the code isn't 100% perfect. I think I have some confusion around some of these myself...

### Issues up for vote vs money on the issues

It strikes me as a little strange that `candidates` have their own app but referendums do not? My reading of this is that the `ballot_measure` app kind of wraps up referendums with the "what are all the things on a voter's ballot" but I see referendums on par with candidates. For example, both have committees which are in support or against and (speculating) have different campaign finance rules governing them. I suppose I see a separate three apps here:
- `ballot` or `election`: everything about what a voter is voting on in a single election, independent of campaign finance.
- `ballot_measure`: everything about individual referendums on a single ballot and how campaign finance relates to them.
- `candidate`: everything about individual candidates/offices up for vote on a single ballot.

Does that make sense? I might have misunderstood the whole `BallotMeasure` model, so maybe some clarification here.

adborden commented:
One more on `zipcode_metro_raw`. If I was to write a management command that reads the zipcode_metro models (or from the db table) and populates `locality` models, would that command live under `locality` or `zipcode_metro_raw`?

adborden commented:
I ran the migrations locally, but ran into this error: https://gist.github.com/adborden/a8c2e1fa8bf6b52264fd

Repro:

```
# on master
$ python manage.py flush # warning: wipes the db
$ python manage.py syncdb
$ python manage.py migrate
# merge PR
$ python manage.py migrate
# error...
```

adborden commented:
Moar praise: Nice naming choice with the Benefactor and Beneficiary. And thanks for updating the Glossary with the proposed terms and term changes.

bcipolli commented:
> One more on zipcode_metro_raw. If I was to write a management command that reads the zipcode_metro models (or from the db table) and populates locality models, would that command live under locality or zipcode_metro_raw?

I would keep `zipcode_metro_raw` as independent of this project, and so make `locality` dependent on `zipcode_metro_raw` and put the code in `locality`.

bcipolli commented:
> Moar praise: Nice naming choice with the Benefactor and Beneficiary. And thanks for updating the Glossary with the proposed terms and term changes.

Thanks! That was a hard one, and chatted with @andrell81 a bit about it before finally getting there. In the code, it works pretty well...

bcipolli commented:
> I ran the migrations locally, but ran into this error

Em... so after chatting with @tdooner , it seemed to make sense that, rather than running migrations, we should just start from scratch. I actually _deleted_ the old migrations, rather than adding the complexity of trying to split the existing fixtures across three apps.

I would gladly enter that data back in, though _some_ of it will no longer need to be entered. We're pretty close to having live data available via NetFile.

bcipolli commented:
@adborden Thanks so much for the feedback and comments. You're probably the most generous and patient reviewer I've encountered on github so far :) I've marked down a few changes. I would like to hold off on the `Election` changes until discussed a bit further--not because I think I'm right, but because changing it takes effort (downstream) and I want to be sure :).

Last comment to address:

> It strikes me as a little strange that candidates have their own app but referendums do not?

that's interesting. I agree that the app structure is asymmetric. 

> I see a separate three apps here:
> - ballot or election: everything about what a voter is voting on in a single election, independent of campaign finance.
> - ballot_measure: everything about individual referendums on a single ballot and how campaign finance relates to them.
> - candidate: everything about individual candidates/offices up for vote on a single ballot.

I think... I agree! Changing from `BallotItem` to `BallotMeasure` actually clarifies a lot. Then `ballot_measure` (which was an abstract proxy for referendum) can become abstract (`ballot` and potentially merged with `election_day`). Then `referendum` can have the same meta-wrapper that `candidate` has. Though perhaps it would be called ... `election`, to indicate the `BallotItem`, not the `BallotItemChoice`.

### My suggestion:

I still feel it's not clear that `election` would contain core ballot information, but instead information about candidates and offices. So my preference is:
- `ballot` - `Ballot`, `BallotItem`, `BallotItemChoice`
- `referendum` - `Referendum` (thin wrapper around `BallotItem`)
- `election` - `Office`,  `Election` (thin wrapper around `BallotItem`), `Candidate` (thin wrapper around `BallotItemChoice`)

The election day event things... I would still want a separate app for that (`election_day`?), as I think there's enough meat, and enough distance from `ballot`, that I'd want to keep it separate.

Lemme know what you think. I will address the non-`election` things first, and wait for more discussion for others.

bcipolli commented:
Hmmm, my thought now is, because `election` is ambiguous, it should be avoided. How about `OfficeElection` and `ElectionDay` models in `office_election` and `election_day` apps, and leave `Election`/`election` behind?

tdooner commented on candidate/models.py:
```diff
@@ -0,0 +1,93 @@
+    # Should alias biography = models.TextField(blank=True)
+
+    def __str__(self):
+        return self.name
```
I like how this normalizes out all the attributes about the person into the `Person` model. I could see that being beneficial also down-the-road to compute stats for the same person across elections.

tdooner commented on candidate/models.py:
```diff
@@ -0,0 +1,93 @@
+        max_length=255, help_text='The party name.')
+
+    def __str__(self):
+        return "The %s party" % self.name
```
A lot of local candidates aren't party-affiliated, as I understand. Glad to see you made the Person's `party_id` foreign key nullable.

tdooner commented on disclosure/settings.py:
```diff
@@ -48,10 +53,10 @@
 REST_FRAMEWORK = {
-    'VIEW_NAME_FUNCTION': ('disclosure_backend.'
+    'VIEW_NAME_FUNCTION': ('disclosure.'
                            'swagger_nickname.view_name')
```
I don't really understand the syntax here, but okay if it works...

adborden replied:
http://django-rest-swagger.readthedocs.org/en/latest/misc.html?highlight=view_name_function#swagger-nickname-attribute

The format itself is the python funciton name. The definition is in `disclosure/swagger_nickname.py` so the module `disclosure.swagger_nickname` contains the function `view_name` so the full "path" is `disclosure.swagger_nickname.view_name` same as you would use if you were `import`ing it.

tdooner commented on candidate/models.py:
```diff
@@ -0,0 +1,93 @@
+            self.office.name, self.locality().dereference())
+
+
+class Candidate(BallotMeasureChoice, SocialMediaModel):
```
A lot of classes inherit from `SocialMediaModel`, which, I suppose, is taking advantage of Python's multiple inheritance feature.

Is there a modular way to mix in methods to classes in python? I only ask because I've recently jumped on the [Composition over Inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance) bandwagon.

bcipolli replied:
:+1: I think this is the concept of "mix-ins", right?

`SocialMediaModel` is supposed to be a mixin. If you want to mix in methods, just put the method on the mixin class... I think that's what you want, right?

adborden replied:
You can use [python decorators](https://pypi.python.org/pypi/decorators) for composition. You can use decorators on functions and classes.

tdooner commented:
No blockers from me. My high-level thoughts are that this is a pretty complex modelling of an election cycle and it will hopefully be sufficient (edge cases nonwithstanding) to get us where we need to go, and further. Do we need this many models to represent the data? Probably not, but I'm up to give it a try.

adborden commented:
> Hmmm, my thought now is, because election is ambiguous, it should be avoided. How about OfficeElection and ElectionDay models in office_election and election_day apps, and leave Election/election behind?

Yeah, I'm down with avoiding Election. I'm not sure what OfficeElection refers to here, "an office up for election" and would potentially inherit/wrap `BallotItem`?

bcipolli commented:
> OfficeElection refers to here, "an office up for election" and would potentially inherit/wrap BallotItem?

Exactly.

Commit:
fix: ballot_measure => ballot

Commit:
fix: candidate => office_election

Commit:
fix: election models.

Commit:
fix: Election => OfficeElection, fix inheritance.

Commit:
fix: Company => Corporation, finance refs.

Commit:
fix: committee_id => filer_id

adborden commented:
Spoke with @bcipolli and @tdooner offline.
- We're in agreement with Ben's suggestion.
- We'll drop the election as a term.
- We're OK to lose the ballot data on production, saving a backup just in case. (saves headaches around data migration).
- The VIP spec is an interesting data point as one way to model ballot information, but a) we wouldn't get any technical benefit from using their terms (only their past experience) b) it's not likely that we would be able to consume their ballot database in the near future.

Commit:
fix: ballot_item_choice => ballot_item_response.

Commit:
fix: office autopopulate BallotItemResponse fields.

Commit:
ENH: add referendum app.

Commit:
RF: make Person model abstract.

Commit:
STY: add "Mixin" suffix to abstract models, make a candidate a Person.

Commit:
STY: remove unused dereference()

Commit:
fix: office

bcipolli commented:
OK, made six more commits to get this somewhere happy. Things aren't perfect, but I am using this in https://github.com/caciviclab/disclosure-backend/pull/124 to auto-populate `Locality`, `Ballot`, `Candidate`, `Referendum`, `IndependentMoney`, and all associated items. The data aren't correct (the parsing I set up is very preliminary), but close enough that I'm confident this schema is usable.

Since yesterday's discussion, I:
- Renamed `BallotItemChoice` to `BallotItemResponse`, to be _ever so slightly_ more in-line with the VIP spec
- Suffixed all abstract models with `Mixin`, to indicate compositional classes.
- Made `Person` an abstract `Mixin`, to avoid a Django limitation on multiple inheritance (and because this isn't a census app; we don't need a `Person` table; `IndividualContributor` and `Candidate` tables will suffice)
- Auto-populate shared `BallotItem`/`BallotItemResponse` fields with those specific to `OfficeElection`/`Referendum` and `Candidate`.

If you want to see details, try checking out the last six commits. I will update https://github.com/caciviclab/disclosure-backend/pull/124 shortly, so you can see where that's at as well.

Commit:
DOC: add help_text, tweak null/default to indicate where auto-set may fail.

Commit:
RF: refactor schema to avoid field duplication, make more appropriate connections.

bcipolli commented:
Scrubbed the models this morning to set `help_text` as well as `null=True` and `default=None` when the fields can fail to be auto-set in the form parsing.

I also refactored a few fields:
- Removed fields from `ElectionDay`, as they were needed, and duplicated on, `Ballot`.
- Moved `party` from `PersonMixin` to `Candidate`. I don't think we want to model political party affiliation of individual contributors.
- Renamed `IndividualBenefactor` to `PersonBenefactor` (clearer, more in-line with other model names)
- Made `Corporation` a mixin (like person), as we aren't trying to model all corporations in the world.
- Moved `ballot_item_response` and `support` off of `IndividualMoney` and onto `Beneficiary`, since it is the `Beneficiary` that must declare its position, and can only stand for one position. @andrell81 is this correct?
- Moved `form` off of `IndividualMoney` and onto `ReportingPeriod`, as `ReportingPeriod` is in reference to specific forms.

I'll be meeting @andrell81 to discuss the schema and parsing (#124) tonight, so perhaps looking to merge this tomorrow would be good?

Commit:
BF: fix up admin. (my bad)

bcipolli commented:
Just reviewed the admin panel. i had broken it; fixed it up now.

It points out that `Address` really should be a mixin (we don't need a master list of addresses, and it'd be nice to edit addresses on the object they refer to). However, somehow that conflicted with locality information... will have to figure that out later, as I have to do real work today lol

Commit:
RF: Make address a Mixin

bcipolli commented:
OK, did the address mixin. Now the admin looks much more reasonable--instead of a dropdown for address with every unique address, you can edit in-place:

![image](https://cloud.githubusercontent.com/assets/4072455/12241195/963251dc-b846-11e5-9987-6045cc973b8d.png)

tdooner commented:
I dig the mixin stuff.

bcipolli commented:
OK, I met with @andrell81 ... no updates imminent here. Does it make sense to merge ahead of tonight and tomorrow's meetings, to get people up to speed on what's changed?

tdooner commented:
Merge merge merge! We can get people up to speed later :)

bcipolli commented:
LOL someone's gotta press the button... it's gonna be me.
